### ISSUE TITLE
Spike: How does clang handle the ARM ABI issues we need to change?

### ISSUE THREAD

**Issue by dmk42** [chapel-lang/chapel#9227]
To make Chapel's LLVM back end compatible with 64-bit ARM, find out how Clang addresses the ABI issues, and figure out how we can do something similar.

**Comment by dmk42** [comment]
Clang has many ABI-related code parts scattered throughout the front end.  Fortunately, most of them deal with C\+\+ related issues like name mangling and non-POD class layout that we don't need to know about.

The Clang ABI code that affects us the most is in `lib/CodeGen/TargetInfo.cpp` which is where the calling sequence code lives.  This is highly complex and special-cased, and it depends on having access to Clang AST nodes.

We could generate Clang AST nodes, but it would be be a large undertaking.  Another option is to copy the logic in Clang but apply it to our generated LLVM code instead of Clang AST.  Then we would have to change the logic whenever Clang added or changed an ABI.

A less severe option would be to generate Chapel procedure calls however we see fit, and only try to match the ABI when we are interfacing with external functions.  For external functions, Clang fills in the needed information for us and we don't have to duplicate all the logic.  The catch is that we then have to tolerate whatever procedure call interface Clang generates.

I did an experiment to see if this might be feasible.  I modified our LLVM code generator to be more tolerant of the Clang calling sequence variations in external functions, rather than trying to enforce the ABI on all functions.  I have had some success so far.  The result passes `release/examples` with `--llvm` with the possible exception of LCALS, which is showing a small numeric difference that may not be significant.  I will have to investigate that later.

Also, with `--llvm --fast --vectorize`, I tried the hellos and they all pass.

It looks like this strategy warrants further work.
